### PR TITLE
Have server return errors in request.

### PR DESCRIPTION
**Pull request by terryxsun**

### CODE REVIEW

File: twitter/src/com/treode/twitter/server/handler/package.scala [hunk 1]
```diff
@@ -30,6 +35,20 @@ package object handler {
       rsp
     }
 
+    def apply (req: Request, notifications: Notification [Unit]): Response = {
+      val rsp = req.response
+      notifications match {
+        case errors @ Errors (_) =>
+          val mapper = new ObjectMapper() with ScalaObjectMapper
```
Comment (Topher-the-Geek) [inline twitter/src/com/treode/twitter/server/handler/package.scala]: Making a mapper is expensive, and a mapper is a fairly static thing. There's one already defined at `com.treode.twitter.finagle.http.mapper` that you should use.

File: twitter/src/com/treode/twitter/server/handler/package.scala [hunk 2]
```diff
@@ -30,6 +35,20 @@ package object handler {
...
+          mapper.registerModule (DefaultScalaModule)
+          rsp.write(mapper.writeValueAsString(errors.strings))
```
Comment (Topher-the-Geek) [inline twitter/src/com/treode/twitter/server/handler/package.scala]: Make a Jackson serializer for Notification. For examples, look at the serializers in com.treode.jackson. You will need to add your serializer to the TreodeModule. Then you can just call `mapper.writeValueAsString (notification)`.

File: core/src/com/treode/notify/Notification.scala [hunk 3]
```diff
@@ -61,11 +62,13 @@ object Notification {
...
-  /** Easily create NoErrors with the given `result`. */
-  def empty [A] (result: A): Notification [A] = NoErrors (result)
-
-  /** Easily create NoErrors with Unit. */
+  /** Easily create a NoErrors with Unit. */
   def empty: Notification [Unit] = NoErrors (())
 
-  def newBuilder: Builder [Unit] = new Builder ()
+  /** Easily build Errors with varargs. */
+  def buildErrors (messages: Message*) : Notification [Unit] =
```
Comment (Topher-the-Geek) [inline core/src/com/treode/notify/Notification.scala]: This is cool. The Scala idiom would be to call this method `apply`. Define it before empty.

File: disk/src/com/treode/disk/messages/package.scala [hunk 4]
```diff
@@ -38,3 +39,9 @@ package messages {
   case class AlreadyDraining (drive: Path) extends Message {
     def en = s"Already draining: ${quote (drive)}"
   }}
+
+package object messages {
+
+  implicit def stringToPath (path: String): Path =
```
Comment (Topher-the-Geek) [inline disk/src/com/treode/disk/messages/package.scala]: One must be very careful with implicit defs. You'll notice that I've kept them largely confined to testing, and to get them you must import a narrow set of test tools. In other words, don't do this here. Instead, you can define it in SpecTools that's a sibling of DrivesAttachHandlerSpec.

Comment (terrynsun) [inline disk/src/com/treode/disk/messages/package.scala]: Where can I define it in SpecTools in order to be able to access it in DrivesAttachHandlerSpec? I tried putting an `implicit def` inside `trait SpecTools` but it doesn't seem to be in scope.

**Comment by Topher-the-Geek** [comment]
You put it in com.treode.twitter.server.handler.SpecTools? DrivesAttachHandlerSpec inherits from that, so the def should be in scope. Can you show me the code?

**Comment by terrynsun** [comment]
I put up a [gist](https://gist.github.com/terrynsun/a7cf86b71668af10dbc6) with what I changed for DrivesAttachHandlerSpec. After adding the `stringToPath` function to SpecTools, I still get a "type mismatch" error from trying to compile the tests.

For the Jackson Serializer, we don't need a Deserializer, right?

I'm also having trouble using a type parameter [A] with a companion object, so that I can write something like `object NotificationSerializer extends StdSerializer [Notification [A]] (classOf [Notification [A]])`.

**Comment by Topher-the-Geek** [comment]
Maybe it's

```
object NotificationSerializer extends StdSerializer [Notification [_]] (classOf [Notification [_]])
```

**Comment by terrynsun** [comment]
That doesn't work for me -- I get the error "Unit is not defined".

**Comment by Topher-the-Geek** [comment]
I just tried it, and had no problems. There must be something else going on in your code.

**Comment by terrynsun** [comment]
You're right -- I just tried it again and it got it to work.

File: core/src/com/treode/notify/Notification.scala [hunk 5]
```diff
@@ -26,12 +26,13 @@ sealed abstract class Notification [+A] {
 
   def isEmpty: Boolean
   def messages: Seq [Message]
+  override def toString : String
```
Comment (Topher-the-Geek) [inline core/src/com/treode/notify/Notification.scala]: This is redundant here. toString is already a method of object, which this inherits from.

File: twitter/src/com/treode/twitter/server/handler/package.scala [hunk 6]
```diff
@@ -30,6 +35,19 @@ package object handler {
       rsp
     }
 
+    def apply (req: Request, notifications: Notification [Unit])
+        (implicit mapper: ObjectMapper): Response = {
```
Comment (Topher-the-Geek) [inline twitter/src/com/treode/twitter/server/handler/package.scala]: There's no need to pass in a mapper. The right mapper is already imported.

File: jackson/src/com/treode/jackson/notification.scala [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+
+package com.treode.jackson
+
+import com.fasterxml.jackson.core.JsonGenerator
+import com.fasterxml.jackson.databind.SerializerProvider
+import com.fasterxml.jackson.databind.ser.std.StdSerializer
+import com.treode.notify.Notification
+
+object NotificationSerializer extends StdSerializer [Notification [_]] (classOf [Notification [_]]) {
```
Comment (Topher-the-Geek) [inline jackson/src/com/treode/jackson/notification.scala]: Also add this serializer in TreodeModule; see the sibling file module.scala.

File: jackson/src/com/treode/jackson/notification.scala [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  def serialize (value: Notification [_], jgen: JsonGenerator, provider: SerializerProvider) {
+    jgen.writeString (value.toString)
```
Comment (Topher-the-Geek) [inline jackson/src/com/treode/jackson/notification.scala]: This should be more like

```
    jgen.writeStartArray()
    for (message <- value.messages)
      codec.writeValue (jgen, message)
    jgen.writeEndArray()
```

To make that work, you also need to define a serializer for Message.

File: twitter/test/com/treode/twitter/server/handler/DrivesAttachHandlerSpec.scala [hunk 9]
```diff
@@ -47,4 +48,21 @@ class DrivesAttachHandlerSpec extends FlatSpec with SpecTools {
...
+  it should "handle errors by returning a JSON array" in
+    served { case (port, controller) =>
+      val notification = Notification(AlreadyAttached("f1"), AlreadyAttached("f2"))
+      (controller.attach _) .expects (Seq.empty) .returning (supply (notification))
+      given
+        .port (port)
+        .body ("[]")
+      .expect
+        .statusCode (400)
+        .body (matchesJson ("""[
```
Comment (Topher-the-Geek) [inline twitter/test/com/treode/twitter/server/handler/DrivesAttachHandlerSpec.scala]: The format should be

```[{message="..."}]```

This gives us a way to add more fields for each error later.

File: jackson/src/com/treode/jackson/notification.scala [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.fasterxml.jackson.core.JsonGenerator
+import com.fasterxml.jackson.databind.SerializerProvider
+import com.fasterxml.jackson.databind.ser.std.StdSerializer
+import com.treode.notify.Notification
+
+object NotificationSerializer extends StdSerializer [Notification [_]] (classOf [Notification [_]]) {
+  def serialize (value: Notification [_], jgen: JsonGenerator, provider: SerializerProvider) {
+    val codec = jgen.getCodec
+    jgen.writeArrayFieldStart("notifications")
```
Comment (Topher-the-Geek) [inline jackson/src/com/treode/jackson/notification.scala]: ArrayFieldStart should only be used when you are certain that you are in an object. Here, you may not be in an object, so use just `writeStartArray`. The tests will need to change.

File: twitter/test/com/treode/twitter/server/handler/DrivesAttachHandlerSpec.scala [hunk 11]
```diff
@@ -47,4 +48,24 @@ class DrivesAttachHandlerSpec extends FlatSpec with SpecTools {
...
+    served { case (port, controller) =>
+      val notification = Notification(AlreadyAttached("f1"), AlreadyAttached("f2"))
+      (controller.attach _) .expects (Seq.empty) .returning (supply (notification))
+      given
+        .port (port)
+        .body ("[]")
+      .expect
+        .statusCode (400)
+        .body (matchesJson ("""
+          "notifications": [
```
Comment (Topher-the-Geek) [inline twitter/test/com/treode/twitter/server/handler/DrivesAttachHandlerSpec.scala]: I'm surprised this worked. It's not valid JSON.

File: jackson/src/com/treode/jackson/message.scala [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+package com.treode.jackson
+
+import com.fasterxml.jackson.core.JsonGenerator
+import com.fasterxml.jackson.databind.SerializerProvider
+import com.fasterxml.jackson.databind.ser.std.StdSerializer
+import com.treode.notify.Message
+
+object MessageSerializer extends StdSerializer [Message] (classOf [Message]) {
+  def serialize (value: Message, jgen: JsonGenerator, provider: SerializerProvider) {
```
Comment (Topher-the-Geek) [inline jackson/src/com/treode/jackson/message.scala]: Notice that `drive.scala` provides many serializers, and so does `cohort.scala`. You can take the same shortcut here; put both Notification and Message into `notify.scala`. Many Scala projects write lots of classes per file for convenience. In large parts of Treode we don't for compilation speed. Here though, that wasn't an issue so we chose convenience. In general, look at sibling files and follow that pattern.

**Pull request closed without merging**